TLM/TLM/Util/ModsCompatibilityChecker.cs

```diff
@@ -1,82 +1,201 @@
-using System;
-using System.Collections.Generic;
-using System.IO;
-using System.Linq;
-using System.Reflection;
-using System.Text;
 using ColossalFramework;
-using ColossalFramework.PlatformServices;
+using ColossalFramework.PlatformServices; // used in RELEASE builds
 using ColossalFramework.Plugins;
 using ColossalFramework.UI;
 using CSUtil.Commons;
+using ICities;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq; // used in RELEASE builds
+using System.Reflection;
 using TrafficManager.UI;
-using UnityEngine;
+using static ColossalFramework.Plugins.PluginManager;
+
+namespace TrafficManager.Util
+{
+    public class ModsCompatibilityChecker
+    {
 
-namespace TrafficManager.Util {
-    public class ModsCompatibilityChecker {
-        //TODO include %APPDATA% mods folder
+        public const ulong LOCAL_MOD = ulong.MaxValue;
 
+        // Used for LoadIncompatibleModsList()
         private const string RESOURCES_PREFIX = "TrafficManager.Resources.";
-        private const string DEFAULT_INCOMPATIBLE_MODS_FILENAME = "incompatible_mods.txt";
-        private readonly ulong[] userModList;
-        private readonly Dictionary<ulong, string> incompatibleModList;
+        private const string INCOMPATIBLE_MODS_FILE = "incompatible_mods.txt";
 
-        public ModsCompatibilityChecker() {
-            incompatibleModList = LoadIncompatibleModList();
-            userModList = GetUserModsList();
+        // parsed contents of incompatible_mods.txt
+        private readonly Dictionary<ulong, string> incompatibleMods;
+
+        public ModsCompatibilityChecker()
+        {
+            incompatibleMods = LoadListOfIncompatibleMods();
         }
 
-        public void PerformModCheck() {
-            Log.Info("Performing incompatible mods check");
-            Dictionary<ulong, string> incompatibleMods = new Dictionary<ulong, string>();
-            for (int i = 0; i < userModList.Length; i++) {
-                string incompatibleModName;
-                if (incompatibleModList.TryGetValue(userModList[i], out incompatibleModName)) {
-                    incompatibleMods.Add(userModList[i], incompatibleModName);
-                }
+        /// <summary>
+        /// Initiates scan for incompatible mods. If any found, and the user has enabled the mod checker, it creates and initialises the modal dialog panel.
+        /// </summary>
+        public void PerformModCheck()
+        {
+            Dictionary<PluginInfo, string> detected = ScanForIncompatibleMods();
+
+            if (detected.Count > 0 && State.GlobalConfig.Instance.Main.ScanForKnownIncompatibleModsAtStartup)
```

Comment:
If `ScanForKnownIncompatibleModsAtStartup == false` we could completely skip the check (no need to call the `ScanForIncompatibleMods` method at all. Will improve performance.

Reply:
The idea here was to scan anyway, and log the results, but only show the dlg if the option is set.

That way if we get user complaining about problems we can look in their log file to see if they are running any known incompatible mods.

Reply:
Added exception catching/logging so if any errors occur it will just silently abort the mod checker.